File: personal/desktop/src-tauri/src/state.rs
//! Wire contract: single-machine-edition design §§4.11, 6 and 9.
//! Nullable observations stay unknown; credentials never enter this document.
use serde::{Deserialize, Serialize};
use std::{
    collections::BTreeMap,
    path::PathBuf,
    sync::{atomic::AtomicBool, RwLock},
};
use tokio::sync::Notify;

#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq)]
pub struct HomeInfo {
    pub path: String,
    #[serde(default)]
    pub version: String,
}
#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq)]
pub struct Docker {
    pub reachable: bool,
}
#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq)]
pub struct Service {
    pub port: Option<u16>,
    pub up: Option<bool>,
}
#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq)]
pub struct Engine {
    pub pid: Option<i32>,
    #[serde(default)]
    pub up: bool,
    pub port: u16,
    pub uptime: Option<f64>,
}
#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq)]
pub struct Queue {
    pub pending: u64,
    pub failed: u64,
    pub last_compile_at: Option<String>,
}
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct SyncConfig {
    #[serde(default = "sync_interval")]
    pub interval_minutes: u64,
    #[serde(default = "sync_enabled")]
    pub enabled: bool,
}
fn sync_interval() -> u64 {
    15
}
fn sync_enabled() -> bool {
    true
}
impl Default for SyncConfig {
    fn default() -> Self {
        Self {
            interval_minutes: 15,
            enabled: true,
        }
    }
}
#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq)]
#[serde(default)]
pub struct SyncResult {
    pub scanned: u64,
    pub new: u64,
    pub increments: u64,
    pub held: u64,
    pub unchanged: u64,
    pub rewritten: u64,
    pub ingested: u64,
    pub skipped: u64,
}
#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq)]
#[serde(default)]
pub struct SyncStatus {
    pub last_run_at: Option<String>,
    pub last_result: Option<SyncResult>,
    pub watching: Vec<String>,
    pub next_due: Option<String>,
    pub next_due_ms: Option<u64>,
    pub running: bool,
    pub held: u64,
}
/// A step is recorded as an ISO timestamp, or derived at read time as a bare `true`
/// (design §5: profile and first compile are observed, never written). Both mean done.
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
#[serde(untagged)]
pub enum StepMark {
    Stamp(String),
    Done(bool),
}
#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq)]
pub struct Steps {
    pub infra: Option<StepMark>,
    pub credentials: Option<StepMark>,
    pub profile: Option<StepMark>,
    pub skill: Option<StepMark>,
    pub first_compile: Option<StepMark>,
}
#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq)]
pub struct LibraryStatus {
    pub name: String,
    #[serde(default)]
    pub current: bool,
    pub engine: Engine,
    pub queue: Option<Queue>,
    pub key: Option<bool>,
    #[serde(default)]
    pub engine_dir: String,
    pub canonical_head: Option<String>,
    pub skill_fresh: Option<bool>,
    #[serde(default)]
    pub steps: Steps,
    pub last_used: Option<String>,
    pub sync: Option<SyncStatus>,
}
#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq)]
pub struct StatusDocument {
    pub home: HomeInfo,
    pub docker: Docker,
    #[serde(default)]
    pub services: BTreeMap<String, Service>,
    #[serde(default)]
    pub libraries: Vec<LibraryStatus>,
}
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct Choices {
    #[serde(default = "default_backend")]
    pub backend: String,
    #[serde(default)]
    pub semantic_retrieval: bool,
    #[serde(default = "default_embedding")]
    pub embedding: String,
    // A library.yaml written before the posture became a choice carries today's behaviour.
    #[serde(default = "default_unattended")]
    pub unattended: bool,
}
fn default_backend() -> String {
    "codex".into()
}
fn default_embedding() -> String {
    "openrouter:openai/text-embedding-3-small".into()
}
fn default_unattended() -> bool {
    true
}
impl Default for Choices {
    fn default() -> Self {
        Self {
            backend: default_backend(),
            semantic_retrieval: false,
            embedding: default_embedding(),
            unattended: default_unattended(),
        }
    }
}
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct ShallowLibrary {
    #[serde(flatten)]
    pub status: LibraryStatus,
    pub tenant: String,
    pub choices: Choices,
    pub pid_alive: bool,
    pub tcp_up: bool,
    #[serde(default)]
    pub watching: Vec<String>,
}
#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq)]
pub struct Shallow {
    pub configured: bool,
    pub home: HomeInfo,
    pub docker: Docker,
    pub services: BTreeMap<String, Service>,
    pub libraries: Vec<ShallowLibrary>,
    pub errors: Vec<String>,
    #[serde(default)]
    pub sync_config: SyncConfig,
}
#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq)]
pub struct Snapshot {
    pub shallow: Shallow,
    pub deep: BTreeMap<String, StatusDocument>,
    /// Unix milliseconds of this observation, including unchanged polls.
    pub fetched_at: u64,
}
impl Snapshot {
    pub fn changed_from(&self, old: &Self) -> bool {
        // Passage of time alone must not trigger a state event. UI clocks derive uptime
        // from fetched_at. All actual health/config/provenance changes still compare.
        let normalize = |mut value: Self| {
            value.fetched_at = 0;
            for lib in &mut value.shallow.libraries {
                lib.status.engine.uptime = None;
            }
            for doc in value.deep.values_mut() {
                for lib in &mut doc.libraries {
                    lib.engine.uptime = None;
                }
            }
            value
        };
        normalize(self.clone()) != normalize(old.clone())
    }
}
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Health {
    Grey,
    Green,
    Amber,
    Red,
}
impl Shallow {
    pub fn health(&self) -> Health {
        if !self.configured {
            return Health::Grey;
        }
        if !self.docker.reachable {
            return Health::Red;
        }
        if !self.errors.is_empty()
            || self.services.len() != 4
            || self.services.values().any(|s| s.up != Some(true))
            || self.libraries.iter().any(|l| !l.status.engine.up)
        {
            Health::Amber
        } else {
            Health::Green
        }
    }
}
pub struct Runtime {
    pub home: PathBuf,
    pub login_path: String,
    pub client: reqwest::Client,
    pub cache: RwLock<Snapshot>,
    pub panel_open: AtomicBool,
    /// When the panel was last ordered on screen (unix ms); a focus loss inside the first
    /// moments after showing is the show itself settling, not the Owner clicking away.
    pub shown_at_ms: std::sync::atomic::AtomicU64,
    pub wants_open: AtomicBool,
    pub ready: AtomicBool,
    pub wake: Notify,
    pub action_lock: tokio::sync::Mutex<()>,
}

#[cfg(test)]
mod tests {
    use super::*;

    fn healthy() -> Snapshot {
        let mut snapshot = Snapshot::default();
        snapshot.shallow.configured = true;
        snapshot.shallow.docker.reachable = true;
        for name in ["postgres", "qdrant", "meili", "rustfs"] {
            snapshot.shallow.services.insert(
                name.into(),
                Service {
                    port: Some(18000),
                    up: Some(true),
                },
            );
        }
        snapshot
    }

    #[test]
    fn colors_need_no_deep_document() {
        assert_eq!(Snapshot::default().shallow.health(), Health::Grey);
        let mut state = healthy();
        assert_eq!(state.shallow.health(), Health::Green);
        state.shallow.services.get_mut("postgres").unwrap().up = Some(false);
        assert_eq!(state.shallow.health(), Health::Amber);
        state.shallow.docker.reachable = false;
        assert_eq!(state.shallow.health(), Health::Red);
    }

    #[test]
    fn elapsed_time_is_not_an_event_but_health_and_queue_changes_are() {
        let mut before = healthy();
        before.deep.insert(
            "notes".into(),
            StatusDocument {
                libraries: vec![LibraryStatus {
                    name: "notes".into(),
                    engine: Engine {
                        port: 18100,
                        up: true,
                        uptime: Some(10.0),
                        ..Default::default()
                    },
                    ..Default::default()
                }],
                ..Default::default()
            },
        );
        let mut after = before.clone();
        after.fetched_at += 5000;
        after.deep.get_mut("notes").unwrap().libraries[0]
            .engine
            .uptime = Some(15.0);
        assert!(!after.changed_from(&before));
        after.deep.get_mut("notes").unwrap().libraries[0].queue = Some(Queue {
            failed: 1,
            ..Default::default()
        });
        assert!(after.changed_from(&before));
    }

    #[test]
    fn deep_response_cannot_forward_undeclared_secrets_to_the_webview() {
        let doc: StatusDocument = serde_json::from_value(serde_json::json!({
            "home": {"path": "/synthetic", "version": "1", "credentials": "synthetic-secret"},
            "docker": {"reachable": true}, "credentials": "synthetic-secret"
        }))
        .unwrap();
        assert!(!serde_json::to_string(&doc)
            .unwrap()
            .contains("synthetic-secret"));
    }
}

File: personal/desktop/src/lib/state.ts
// Copied wire types from the single-machine-edition design §§4.11, 6, 9 and 10.
// This client has no runtime/source dependency on the framework or console projects.
export type Known = boolean | null;
export type Backend = 'codex' | 'claude-code' | 'api';
export type Tab = 'dashboard' | 'search' | 'settings';
export interface Engine { pid: number | null; up: boolean; port: number; uptime: number | null }
export type StepMark = string | boolean | null;
export interface Steps {
  infra: StepMark; credentials: StepMark; profile: StepMark; skill: StepMark; first_compile: StepMark;
}
export const stepLabels: [keyof Steps, string][] = [
  ['infra', 'Infra'], ['credentials', 'Key'], ['profile', 'Profile'],
  ['skill', 'Skill'], ['first_compile', 'Compile'],
];
export interface SyncStatus {
  last_run_at: string | null;
  last_result: { scanned: number; new: number; increments: number; held: number; unchanged: number; rewritten: number; ingested: number; skipped: number } | null;
  watching: string[]; next_due: string | null; running: boolean; held: number;
}
export interface LibraryStatus {
  name: string; current: boolean; engine: Engine;
  queue: { pending: number; failed: number; succeeded?: number; failed_by_kind?: Record<string, number>; last_compile_at: string | null } | null;
  key: Known; engine_dir: string; canonical_head: string | null;
  skill_fresh: Known; steps: Steps; last_used: string | null;
  sync?: SyncStatus | null;
}
export interface Choices { backend: Backend; semantic_retrieval: boolean; embedding: string; unattended: boolean }
export interface ShallowLibrary extends LibraryStatus { tenant: string; choices: Choices; pid_alive: boolean; tcp_up: boolean; watching?: string[] }
export interface StatusDocument {
  home: { path: string; version: string }; docker: { reachable: boolean };
  services: Record<string, { port: number | null; up: Known }>;
  libraries: LibraryStatus[];
}
export interface Shallow extends Omit<StatusDocument, 'libraries'> {
  configured: boolean; libraries: ShallowLibrary[]; errors: string[];
  sync_config?: { interval_minutes: number; enabled: boolean };
}
export interface Snapshot { shallow: Shallow; deep: Record<string, StatusDocument>; fetched_at: number }
export interface Claim {
  anchor: string; document_path: string; text: string;
  citations: { source_id: string; block_start: number; block_end: number }[];
}
export interface SourceWindow { source_id: string; block_start: number; block_end: number; text: string }
export interface RecallResult {
  answer: string; used_claims: Claim[]; used_windows: SourceWindow[];
  citation_handles: Record<string, string>; documents_read: string[];
  used_episode_summaries?: SourceWindow[];
  used_component_evidence?: { claims: Claim[]; windows: SourceWindow[] }[];
  // Resolved by the desktop's read-only API client, because console routes use IDs.
  document_ids: Record<string, string>;
}
export const emptyState: Snapshot = {
  shallow: { configured: false, home: { path: '~/.pkc', version: '' }, docker: { reachable: false }, services: {}, libraries: [], errors: [] },
  deep: {}, fetched_at: 0,
};
export function health(state: Shallow): 'grey' | 'green' | 'amber' | 'red' {
  if (!state.configured) return 'grey';
  if (!state.docker.reachable) return 'red';
  if (state.errors.length || Object.keys(state.services).length !== 4
    || Object.values(state.services).some(s => s.up !== true) || state.libraries.some(l => !l.engine.up)) return 'amber';
  return 'green';
}
export const healthLabels = { grey: 'Set up your home', green: 'All engines ready', amber: 'Needs attention', red: 'Docker is unreachable' };
export function currentLibrary(state: Snapshot): ShallowLibrary | undefined {
  return state.shallow.libraries.find(l => l.current);
}
/**
 * What the pane may say about the stored embedding key, from observation alone.
 * Engine startup is fail-closed — with semantic retrieval on, the engine probes the
 * provider with the stored key and refuses to come up if it is rejected — so a running
 * engine is proof the provider accepted it. A stored key nothing has exercised is
 * reported as stored and no more. Without a library nothing has looked, and the pane
 * must stay usable, so the empty state (the fill-in field) is what the Owner sees.
 */
export type KeyReadout = 'absent' | 'unknown' | 'verified' | 'unverified' | 'unused';
export function keyReadout(library: ShallowLibrary | undefined, deep?: LibraryStatus): KeyReadout {
  if (!library) return 'absent';
  // Only the engine's own document knows whether a credential is present; the disk walk
  // never looks. Same overlay rule as every other readout: the deep answer, or nothing.
  const key = deep?.key ?? library.key;
  if (key === false) return 'absent';
  if (key === null) return 'unknown';
  if (!library.engine.up) return 'unverified';
  return library.choices.semantic_retrieval ? 'verified' : 'unused';
}
/** The credential the library's embedding provider is asked for. */
export function credentialName(embedding: string | undefined): string {
  const provider = embedding?.split(':')[0];
  if (provider === 'openai') return 'OPENAI_API_KEY';
  if (provider === 'google' || provider === 'google-genai') return 'GOOGLE_API_KEY';
  return 'OPENROUTER_API_KEY';
}
export function deepLibrary(state: Snapshot, library: ShallowLibrary): LibraryStatus | undefined {
  // Only the selected engine's own answer supplies its deep overlay. A sibling's
  // cached status cannot make a stopped engine appear healthy.
  return library.tcp_up ? state.deep[library.name]?.libraries.find(l => l.name === library.name) : undefined;
}
export function syncSummary(sync: SyncStatus | null | undefined, watching: string[] = []): string {
  return `watching ${watching.length} dirs · last sync ${sync?.last_run_at ? timeLabel(sync.last_run_at) : 'never'} · held ${sync?.held ?? '—'}`;
}
export function consoleUrl(port: number, kind?: 'document' | 'claim' | 'source', id?: string, anchor?: string | number): string {
  const root = `http://127.0.0.1:${port}/`;
  if (!kind || !id) return root;
  const view = kind === 'source' ? 'sources' : 'library';
  return root + '#/' + [view, kind, id, ...(anchor != null ? [String(anchor)] : [])].map(encodeURIComponent).join('/');
}
export function uptime(seconds: number | null): string {
  if (seconds === null) return 'Uptime unknown';
  if (seconds < 60) return 'Just started';
  const minutes = Math.floor(seconds / 60);
  if (minutes < 60) return `${minutes}m uptime`;
  const hours = Math.floor(minutes / 60);
  return `${hours}h ${minutes % 60}m uptime`;
}
export function timeLabel(stamp: string | boolean | null | undefined): string {
  if (!stamp) return 'Unknown';
  if (stamp === true) return 'done';
  const date = new Date(stamp);
  return Number.isNaN(date.valueOf()) ? 'Unknown' : date.toLocaleString([], { month: 'short', day: 'numeric', hour: '2-digit', minute: '2-digit' });
}
